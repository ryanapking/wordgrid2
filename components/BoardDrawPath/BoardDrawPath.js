import React from 'react';
import { View } from 'react-native';
import PropTypes from 'prop-types';

const BoardDrawPath = props => {
  const { square1, square2, boardLocation } = props;

  const height = boardLocation.rowHeight * .1;
  const borderRadius = height / 2;

  const p1 = getRelativeCoordinates(boardLocation, square1.rowIndex, square1.columnIndex);
  const p2 = getRelativeCoordinates(boardLocation, square2.rowIndex, square2.columnIndex);
  const angle = calculateAngle(p1, p2);
  const length = calculateLength(p1, p2) + borderRadius;
  const midpoint = calculateMidPoint(p1, p2);

  const style = {
    left: midpoint.x - (length / 2),
    top: midpoint.y - (height / 2),
    width: length,
    transform: [{rotate: angle + 'deg'}],
    height: height,
    borderRadius: borderRadius,
    backgroundColor: '#5283ff',
    position: 'absolute'
  };

  return (
    <View style={style} />
  );
};

BoardDrawPath.propTypes = {
  square1: PropTypes.shape({
    rowIndex: PropTypes.number,
    columnIndex: PropTypes.number,
  }),
  square2: PropTypes.shape({
    rowIndex: PropTypes.number,
    columnIndex: PropTypes.number,
  }),
  boardLocation: PropTypes.shape({
    rowHeight: PropTypes.number,
    columnWidth: PropTypes.number,
  }),
};

const getRelativeCoordinates = (boardLocation, rowIndex, columnIndex) => {
  const y = (rowIndex + .5) * boardLocation.rowHeight;
  const x = (columnIndex + .5) * boardLocation.columnWidth;
  return { x, y };
};

const calculateLength = (p1, p2) => {
  return Math.hypot(p1.x - p2.x, p1.y - p2.y);
};

const calculateMidPoint = (p1, p2) => {
  const x = (p1.x + p2.x) / 2;
  const y = (p1.y + p2.y) / 2;
  return {x, y};
};

const calculateAngle = (p1, p2) => {
  let dy = p1.y - p2.y;
  let dx = p1.x - p2.x;
  let theta = Math.atan2(dy, dx); // range (-PI, PI]
  theta *= 180 / Math.PI; // rads to degs, range (-180, 180]
  //if (theta < 0) theta = 360 + theta; // range [0, 360)
  return theta;
};

export default BoardDrawPath;