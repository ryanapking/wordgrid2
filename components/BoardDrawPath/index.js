/* istanbul ignore file */
export { default } from "./BoardDrawPath";