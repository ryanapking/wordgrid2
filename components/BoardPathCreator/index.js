/* istanbul ignore file */
export { default } from './BoardPathCreator';