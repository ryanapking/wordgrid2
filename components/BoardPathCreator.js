import React from 'react';
import { View, StyleSheet } from 'react-native';
import PropTypes from "prop-types";

import BoardDrawPath from './BoardDrawPath';

const BoardPathCreator = props => {
  const { squares, boardLocation } = props;

  if (!boardLocation.rowHeight || !boardLocation.columnWidth) return null;

  // split the consumedSquares into pairs, to allow a path section to be rendered
  const squarePairs = squares
    .map( (square, squareIndex) => {
      return squares.slice(squareIndex - 1, squareIndex + 1);
    })
    .filter( (pair) => {
      return pair.length === 2;
    });

  return (
    <View style={styles.overlay} pointerEvents={'none'}>
      {squarePairs.map( (pair, pairIndex) =>
        <BoardDrawPath key={pairIndex} square1={pair[0]} square2={pair[1]} boardLocation={boardLocation}/>
      )}
    </View>
  );
};

BoardPathCreator.propTypes = {
  squares: PropTypes.arrayOf(
    PropTypes.shape({
      rowIndex: PropTypes.number,
      columnIndex: PropTypes.number,
    })
  ),
  boardLocation: PropTypes.shape({
    rowHeight: PropTypes.number,
    columnWidth: PropTypes.number,
  }),
};


const styles = StyleSheet.create({
  overlay: {
    width: '100%',
    height: '100%',
    position: 'absolute'
  }
});

export default BoardPathCreator;