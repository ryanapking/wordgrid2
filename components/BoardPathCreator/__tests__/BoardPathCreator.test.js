import React from 'react';
import { render } from 'react-native-testing-library';

import BoardPathCreator from "..";
import BoardDrawPath from "../../BoardDrawPath";

// mock BoardDrawPath to reduce snapshot size
jest.mock('../../BoardDrawPath', () => 'BoardDrawPath');

const testProps = {
  squares: [
    {rowIndex: 6, columnIndex: 4},
    {rowIndex: 6, columnIndex: 3},
    {rowIndex: 5, columnIndex: 3},
    {rowIndex: 5, columnIndex: 2},
    {rowIndex: 4, columnIndex: 2},
    {rowIndex: 3, columnIndex: 3},
  ],
  boardLocation: {rowHeight: 38.6, columnWidth: 37.5},
};

describe('BoardPathCreator', () => {
  it('creates the correct number of paths for the given squares', () => {
    const { unmount, queryAllByType } = render(
      <BoardPathCreator boardLocation={testProps.boardLocation} squares={testProps.squares} />
    );
    expect(queryAllByType(BoardDrawPath)).toHaveLength(testProps.squares.length - 1);
    unmount();
  });

  it('renders as expected', () => {
    const { unmount, toJSON } = render(
      <BoardPathCreator boardLocation={testProps.boardLocation} squares={testProps.squares} />
    );
    expect(toJSON()).toMatchSnapshot();
    unmount();
  });

  it('renders null without boardLocation', () => {
    const { unmount, toJSON } = render(
      <BoardPathCreator squares={testProps.squares} />
    );
    expect(toJSON()).toBeNull();
    unmount();
  });
});