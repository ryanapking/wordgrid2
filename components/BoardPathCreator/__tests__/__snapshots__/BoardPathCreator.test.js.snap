// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardPathCreator renders as expected 1`] = `
<View
  pointerEvents="none"
  style={
    Object {
      "height": "100%",
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <BoardDrawPath
    boardLocation={
      Object {
        "columnWidth": 37.5,
        "rowHeight": 38.6,
      }
    }
    square1={
      Object {
        "columnIndex": 4,
        "rowIndex": 6,
      }
    }
    square2={
      Object {
        "columnIndex": 3,
        "rowIndex": 6,
      }
    }
  />
  <BoardDrawPath
    boardLocation={
      Object {
        "columnWidth": 37.5,
        "rowHeight": 38.6,
      }
    }
    square1={
      Object {
        "columnIndex": 3,
        "rowIndex": 6,
      }
    }
    square2={
      Object {
        "columnIndex": 3,
        "rowIndex": 5,
      }
    }
  />
  <BoardDrawPath
    boardLocation={
      Object {
        "columnWidth": 37.5,
        "rowHeight": 38.6,
      }
    }
    square1={
      Object {
        "columnIndex": 3,
        "rowIndex": 5,
      }
    }
    square2={
      Object {
        "columnIndex": 2,
        "rowIndex": 5,
      }
    }
  />
  <BoardDrawPath
    boardLocation={
      Object {
        "columnWidth": 37.5,
        "rowHeight": 38.6,
      }
    }
    square1={
      Object {
        "columnIndex": 2,
        "rowIndex": 5,
      }
    }
    square2={
      Object {
        "columnIndex": 2,
        "rowIndex": 4,
      }
    }
  />
  <BoardDrawPath
    boardLocation={
      Object {
        "columnWidth": 37.5,
        "rowHeight": 38.6,
      }
    }
    square1={
      Object {
        "columnIndex": 2,
        "rowIndex": 4,
      }
    }
    square2={
      Object {
        "columnIndex": 3,
        "rowIndex": 3,
      }
    }
  />
</View>
`;
